src/theory/strings/theory_strings.cpp

```diff
@@ -4110,51 +4088,57 @@ Node TheoryStrings::mkLength( Node t ) {
   return Rewriter::rewrite( NodeManager::currentNM()->mkNode( kind::STRING_LENGTH, t ) );
 }
 
-Node TheoryStrings::mkExplain( std::vector< Node >& a ) {
+Node TheoryStrings::mkExplain(const std::vector<Node>& a)
+{
   std::vector< Node > an;
   return mkExplain( a, an );
 }
 
-Node TheoryStrings::mkExplain( std::vector< Node >& a, std::vector< Node >& an ) {
+Node TheoryStrings::mkExplain(const std::vector<Node>& a,
+                              const std::vector<Node>& an)
+{
   std::vector< TNode > antec_exp;
-  for( unsigned i=0; i<a.size(); i++ ) {
-    if( std::find( a.begin(), a.begin() + i, a[i] )==a.begin() + i ) {
-      bool exp = true;
-      Debug("strings-explain") << "Ask for explanation of " << a[i] << std::endl;
-      //assert
-      if(a[i].getKind() == kind::EQUAL) {
-        //Assert( hasTerm(a[i][0]) );
-        //Assert( hasTerm(a[i][1]) );
-        Assert( areEqual(a[i][0], a[i][1]) );
-        if( a[i][0]==a[i][1] ){
-          exp = false;
-        }
-      } else if( a[i].getKind()==kind::NOT && a[i][0].getKind()==kind::EQUAL ) {
-        Assert( hasTerm(a[i][0][0]) );
-        Assert( hasTerm(a[i][0][1]) );
-        AlwaysAssert( d_equalityEngine.areDisequal(a[i][0][0], a[i][0][1], true) );
-      }else if( a[i].getKind() == kind::AND ){
-        for( unsigned j=0; j<a[i].getNumChildren(); j++ ){
-          a.push_back( a[i][j] );
-        }
-        exp = false;
-      }
-      if( exp ) {
-        unsigned ps = antec_exp.size();
-        explain(a[i], antec_exp);
-        Debug("strings-explain") << "Done, explanation was : " << std::endl;
-        for( unsigned j=ps; j<antec_exp.size(); j++ ) {
-          Debug("strings-explain") << "  " << antec_exp[j] << std::endl;
-        }
-        Debug("strings-explain") << std::endl;
+  // copy to processing vector
+  std::vector<Node> ap;
+  ap.insert(ap.end(), a.begin(), a.end());
+  for (unsigned i = 0; i < ap.size(); i++)
+  {
+    Node api = ap[i];
+    if (std::find(ap.begin(), ap.begin() + i, api) != ap.begin() + i)
+    {
+      // already processed
+      continue;
+    }
+    if (api.getKind() == AND)
+    {
+      for (const Node& apic : api)
+      {
+        ap.push_back(apic);
       }
+      continue;
+    }
+    Debug("strings-explain") << "Add to explanation " << api << std::endl;
+    if (api.getKind() == NOT && api[0].getKind() == EQUAL)
+    {
+      Assert(hasTerm(api[0][0]));
+      Assert(hasTerm(api[0][1]));
+      // ensure that we are ready to explain the disequality
+      AlwaysAssert(d_equalityEngine.areDisequal(api[0][0], api[0][1], true));
     }
+    Assert(api.getKind() != EQUAL || d_equalityEngine.areEqual(api[0], api[1]));
```

Comment:
One detail that is different here is that we are not skipping cases where `api[0] == api[1]`. Is this change intentional?

Reply:
Yes, I realized that the core `explain` method is robust to this case, so this was being checked twice.